File: Lampac/Modules/OnlineRUS/Zetflix/Service.cs
using Shared.Models.Base;
using Shared.Models.Templates;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Web;

namespace Zetflix
{
    public struct ZetflixInvoke
    {
        string host, apihost;
        bool usehls;
        Func<string, string> onstreamfile;
        Func<string, List<HeadersModel>, Task<string>> onget;

        public ZetflixInvoke(string host, string apihost, bool hls, Func<string, List<HeadersModel>, Task<string>> onget, Func<string, string> onstreamfile)
        {
            this.host = host != null ? $"{host}/" : null;
            this.apihost = apihost;
            this.onstreamfile = onstreamfile;
            this.onget = onget;
            usehls = hls;
        }

        public async Task<EmbedModel> Embed(long kinopoisk_id, int s)
        {
            string html = await onget.Invoke($"{apihost}/iplayer/videodb.php?kp={kinopoisk_id}" + (s > 0 ? $"&season={s}" : ""), HeadersModel.Init(
                ("dnt", "1"),
                ("pragma", "no-cache"),
                ("referer", "https://www.google.com/"),
                ("upgrade-insecure-requests", "1")
            ));

            return Embed(html);
        }

        public EmbedModel Embed(string html)
        {
            if (html == null)
                return null;

            string quality = html.Contains("1080p") ? "1080p" : html.Contains("720p") ? "720p" : "480p";
            string check_url = Regex.Match(html, "(https?://[^\\[\\|,\n\r\t ]+\\.mp4)").Groups[1].Value;

            string file = Regex.Match(html, "file:(\\[[^\n\r]+\\])(,|}\\) ;)").Groups[1].Value;
            if (string.IsNullOrWhiteSpace(file))
                file = Regex.Match(html, "file:(\\[[\\s\\S]+?\\])\\s*(,|\\}\\) ;)", RegexOptions.Multiline).Groups[1].Value;

            if (string.IsNullOrWhiteSpace(file))
            {
                file = Regex.Match(html, "file:\"([^\"]+)\"").Groups[1].Value;
                if (!string.IsNullOrWhiteSpace(file))
                {
                    return new EmbedModel()
                    {
                        pl = new List<RootObject>()
                        {
                            new RootObject() { file = file, title = "Дубляж" }
                        },
                        movie = true,
                        quality = quality,
                        check_url = check_url
                    };
                }

                return null;
            }

            file = Regex.Replace(file.Trim(), "(\\{|, )([a-zA-Z_][a-zA-Z0-9_]*): ?", "$1\"$2\":")
                        .Replace("},]", "}]");

            List<RootObject> pl = null;

            try
            {
                pl = JsonSerializer.Deserialize<List<RootObject>>(file);
                if (pl == null || pl.Count == 0)
                    return null;
            }
            catch
            {
                return null;
            }

            bool isMovie = !file.Contains("\"comment\":", StringComparison.OrdinalIgnoreCase);
            return new EmbedModel() { pl = pl, movie = isMovie, quality = quality, check_url = check_url };
        }

        public async Task<int> number_of_seasons(long id)
        {
            int number_of_seasons = 1;
            string themoviedb = await onget.Invoke($"https://api.themoviedb.org/3/tv/{id}?api_key=4ef0d7355d9ffb5151e987764708ce96&language=ru-RU", HeadersModel.Init(
                ("accept", "application/json")
            ));

            if (themoviedb != null)
            {
                try
                {
                    var root = JsonSerializer.Deserialize<JsonObject>(themoviedb);
                    number_of_seasons = root["number_of_seasons"].GetValue<int>();
                    if (number_of_seasons < 1)
                        number_of_seasons = 1;
                }
                catch { }
            }

            if (number_of_seasons <= 0)
                number_of_seasons = 1;

            return number_of_seasons;
        }

        public ITplResult Tpl(EmbedModel root, int number_of_seasons, long id, long kinopoisk_id, string title, string original_title, string t, int s, bool isbwa = false, bool rjson = false, VastConf vast = null)
        {
            if (root?.pl == null || root.pl.Count == 0)
                return default;

            if (root.movie)
            {
                var mtpl = new MovieTpl(title, original_title, root.pl.Count);

                foreach (var pl in root.pl)
                {
                    string name = pl.title;
                    string file = pl.file;

                    if (string.IsNullOrWhiteSpace(name) || string.IsNullOrWhiteSpace(file))
                        continue;

                    var streamquality = new StreamQualityTpl();

                    foreach (Match m in Regex.Matches(file, "\\[(1080|720|480|360)p?\\]([^\\[\\|,\n\r\t ]+\\.(mp4|m3u8))"))
                    {
                        string link = m.Groups[2].Value;
                        if (string.IsNullOrEmpty(link))
                            continue;

                        if (usehls && !link.Contains(".m3u"))
                            link += ":hls:manifest.m3u8";
                        else if (!usehls && link.Contains(".m3u"))
                            link = link.Replace(":hls:manifest.m3u8", "");

                        if (isbwa)
                            link = Regex.Replace(link, "/([0-9]+)\\.(m3u8|mp4)", $"/{m.Groups[1].Value}.$2");

                        streamquality.Insert(onstreamfile.Invoke(link), $"{m.Groups[1].Value}p");
                    }

                    mtpl.Append(name, streamquality.Firts().link, streamquality: streamquality, vast: vast);
                }

                return mtpl;
            }

            string enc_title = HttpUtility.UrlEncode(title);
            string enc_original_title = HttpUtility.UrlEncode(original_title);

            if (s == -1)
            {
                var tpl = new SeasonTpl(root.quality);

                for (int i = 1; i <= number_of_seasons; i++)
                {
                    string link = host + $"lite/zetflix?serial=1&id={id}&rjson={rjson}&kinopoisk_id={kinopoisk_id}&title={enc_title}&original_title={enc_original_title}&s={i}";
                    tpl.Append($"{i} сезон", link, i);
                }

                return tpl;
            }

            var vtpl = new VoiceTpl();
            var etpl = new EpisodeTpl();
            var hashvoices = new HashSet<string>();
            string sArch = s.ToString();

            foreach (var episode in root.pl.AsEnumerable().Reverse())
            {
                var episodes = episode.folder;
                if (episodes == null || episodes.Length == 0)
                    continue;

                string perevod = episode.title;
                if (perevod != null && string.IsNullOrEmpty(t))
                    t = perevod;

                if (perevod != null && !hashvoices.Contains(perevod))
                {
                    hashvoices.Add(perevod);
                    string link = host + $"lite/zetflix?serial=1&id={id}&rjson={rjson}&kinopoisk_id={kinopoisk_id}&title={enc_title}&original_title={enc_original_title}&s={s}&t={HttpUtility.UrlEncode(perevod)}";

                    vtpl.Append(perevod, t == perevod, link);
                }

                if (perevod != t)
                    continue;

                foreach (var pl in episodes)
                {
                    string name = pl.comment;
                    string episodeNum = Regex.Match(name ?? string.Empty, "^([0-9]+)").Groups[1].Value;

                    if (string.IsNullOrWhiteSpace(name) || string.IsNullOrWhiteSpace(episodeNum))
                        continue;

                    string link = host + $"lite/zetflix/video?serial=1&kinopoisk_id={kinopoisk_id}&title={enc_title}&original_title={enc_original_title}&s={s}&e={episodeNum}&t={HttpUtility.UrlEncode(perevod)}";
                    etpl.Append(name, title ?? original_title, sArch, episodeNum, link, "call", streamlink: link + "&play=true", voice_name: perevod, vast: vast);
                }
            }

            etpl.Append(vtpl);

            return etpl;
        }
    }
}

File: Lampac/examples/modules/KinoGramJS/service.js
async function handle() {
    var q = req.query || {};

    var imdb_id = q.imdb_id || "";
    var kinopoisk_id = toInt(q.kinopoisk_id, 0);
    var title = q.title || "";
    var original_title = q.original_title || "";
    var year = toInt(q.year, 0);
    var serial = toInt(q.serial, 0);
    var t = q.t || "";
    var s = toInt(q.s, -1);

    var result = await search(imdb_id, kinopoisk_id, serial);
    if (result == null)
        return "";

    if (result.movie != null)
    {
        var tpl = new MovieTpl(title, original_title);

        for (var i = 0; i < result.movie.length; i++)
        {
            var movie = result.movie[i];
            var streamquality = new StreamQualityTpl();

            for (var j = 0; j < movie.links.length; j++) {
                var l = movie.links[j];
                streamquality.Append(streamProxy(l.link), l.quality);
            }

            var first = streamquality.Firts();

            tpl.Append(
                movie.translation,
                first.link,
                "play",
                null,
                streamquality,
                null,
                null,
                null,
                null,
                first.quality
            );
        }

        return tpl.ToHtml();
    }
    else
    {
        if (result.serial == null)
            return "";

        var defaultargs =
            "&imdb_id=" + encodeURIComponent(imdb_id) +
            "&kinopoisk_id=" + kinopoisk_id +
            "&title=" + encodeURIComponent(title) +
            "&original_title=" + encodeURIComponent(original_title) +
            "&serial=" + serial;

        if (s === -1)
        {
            var tpl = new SeasonTpl("4K HDR");

            for (var seasonKey in result.serial)
            {
                tpl.Append(
                    seasonKey + " сезон",
                    host + "/lite/kinogram?s=" + seasonKey + defaultargs,
                    parseInt(seasonKey, 10)
                );
            }

            return tpl.ToHtml();
        }
        else
        {
            var vtpl = new VoiceTpl();
            var tpl = new EpisodeTpl();

            var seasonVoices = result.serial[String(s)];
            if (seasonVoices == null || seasonVoices.length === 0)
                return "";

            var activTranslate = t;
            for (var k = 0; k < seasonVoices.length; k++)
            {
                var translation = seasonVoices[k];

                if (!activTranslate)
                    activTranslate = translation.id;

                vtpl.Append(
                    translation.name,
                    activTranslate === translation.id,
                    host + "/lite/kinogram?s=" + s + "&t=" + translation.id + defaultargs
                );
            }

            var activeVoice = null;
            for (var n = 0; n < seasonVoices.length; n++)
            {
                if (seasonVoices[n].id === activTranslate) {
                    activeVoice = seasonVoices[n];
                    break;
                }
            }

            if (activeVoice == null)
                return "";

            for (var m = 0; m < activeVoice.episodes.length; m++)
            {
                var episode = activeVoice.episodes[m];
                var streamquality = new StreamQualityTpl();

                for (var z = 0; z < episode.links.length; z++) {
                    var l = episode.links[z];
                    streamquality.Append(streamProxy(l.link), l.quality);
                }

                var first = streamquality.Firts();

                tpl.Append(
                    episode.id + " серия",
                    title || original_title,
                    String(s),
                    episode.id,
                    first.link,
                    "play",
                    streamquality
                );
            }

            tpl.Append(vtpl);

            return tpl.ToHtml();
        }
    }
}


async function search(imdb_id, kinopoisk_id, serial)
{
    /*const cachekey = `${imdb_id}:${kinopoisk_id}:${serial}`;
    var html = cacheGet(cachekey);
    if (!html) {
        html = await httpGet("https://kinogram.com/movie?imdb_id=" + encodeURIComponent(imdb_id) + "&kinopoisk_id=" + kinopoisk_id);
        cacheSet(cachekey, html, 20);
    }*/

    // html parse ...

    var defaultLinks = [
        { link: "https://www.elecard.com/storage/video/TheaterSquare_3840x2160.mp4", quality: "2160p" },
        { link: "https://www.elecard.com/storage/video/TheaterSquare_1920x1080.mp4", quality: "1080p" },
        { link: "https://www.elecard.com/storage/video/TheaterSquare_1280x720.mp4", quality: "720p" }
    ];

    var res = null;

    if (serial === 0)
    {
        res = {
            movie: [
                {
                    translation: "RHS",
                    links: defaultLinks
                },
                {
                    translation: "ViruseProject",
                    links: defaultLinks
                }
            ]
        };
    }
    else
    {
        res = {
            serial: {
                "1": [
                    {
                        id: "36",
                        name: "ViruseProject",
                        episodes: [
                            { id: "1", links: defaultLinks },
                            { id: "2", links: defaultLinks },
                            { id: "3", links: defaultLinks }
                        ]
                    },
                    {
                        id: "12",
                        name: "RHS",
                        episodes: [
                            { id: "1", links: defaultLinks },
                            { id: "2", links: defaultLinks }
                        ]
                    }
                ],
                "2": [
                    {
                        id: "36",
                        name: "ViruseProject",
                        episodes: [
                            { id: "1", links: defaultLinks }
                        ]
                    },
                    {
                        id: "12",
                        name: "RHS",
                        episodes: [
                            { id: "1", links: defaultLinks }
                        ]
                    }
                ]
            }
        };
    }

    return res;
}


function toInt(v, def = 0) {
    var n = parseInt(v, 10);
    return isNaN(n) ? def : n;
}
